src/services/assistant.ts

```diff
@@ -0,0 +1,236 @@
+        } catch (_) {}
+        return out
+      }
+
+      // append assistant tool call message (not shown in UI)
+      const assistantCallMessage: ChatMessage = {
+        role: 'assistant',
+        content: '',
+        // @ts-ignore
+        tool_calls: toolCalls,
+      }
+      convo = [...convo, assistantCallMessage]
+
+      // execute ONLY the first tool call (log name and args for debugging)
+      const call = toolCalls[0]
+      const name = call.function?.name
+      let args: any = {}
+      try { args = JSON.parse(call.function?.arguments || '{}') } catch (_) {}
+      console.log('[assistant][tool-call]', name, args)
+
+      // Special handling: suggest a clickable Google search instead of executing dummy search_engine
+      if (name === 'search_engine') {
+        const q = typeof args?.query === 'string' ? args.query : ''
+        const url = `https://www.google.com/search?q=${encodeURIComponent(q)}`
+        // Return a special marker for the UI to render a button
+        try {
+          const usages = assistantStore.getTokenUsages()
+          const totals = usages.reduce(
+            (acc, u) => ({
+              prompt: acc.prompt + (u.prompt || 0),
+              completion: acc.completion + (u.completion || 0),
+              total: acc.total + (u.total || 0),
+            }),
+            { prompt: 0, completion: 0, total: 0 },
+          )
+          console.log('[assistant][tokens] prompt=%d completion=%d total=%d', totals.prompt, totals.completion, totals.total)
+        } catch (_) {}
+        return `SEARCH_BUTTON:${url}|Search the web for \"${q}\"`
+      }
+      const fn = TOOL_DISPATCH[name]
+      const output = await (fn ? fn(args) : Promise.resolve({ error: `Unknown tool: ${name}` }))
+      const toolMessage: ChatMessage = {
+        role: 'tool',
+        content: typeof output === 'string' ? output : JSON.stringify(output),
+        // @ts-ignore
+        tool_call_id: call.id,
+      }
+      convo = [...convo, toolMessage]
+
+      iterations += 1
+      if (iterations > 20) {
+        console.warn('[assistant] breaking after too many tool iterations')
+        break
+      }
+    }
+
+    try {
+      const usages = assistantStore.getTokenUsages()
+      const totals = usages.reduce(
+        (acc, u) => ({
+          prompt: acc.prompt + (u.prompt || 0),
+          completion: acc.completion + (u.completion || 0),
+          total: acc.total + (u.total || 0),
+        }),
+        { prompt: 0, completion: 0, total: 0 },
+      )
+      console.log('[assistant][tokens] prompt=%d completion=%d total=%d', totals.prompt, totals.completion, totals.total)
+    } catch (_) {}
+    return 'No content returned.'
+  }
+}
+
+export const assistantStore = AssistantStore.getInstance()
+export const assistantService = new AssistantService()
+// Eagerly initialize engine (non-blocking)
+void assistantService.initialize()
```

Comment:
this looks a bit strange as this will eagerly initialize when loading the file. you do initialize on each send too